tt_metal/hw/inc/dataflow_api.h

```diff
@@ -1537,40 +1561,143 @@ class Buffer {
+        this->init(type, bank_base_address, page_size);
+    }
+
+    Buffer(uint32_t page_size, uint32_t num_cores,  uint32_t addr, volatile tt_l1_ptr uint32_t* command_ptr) {
+        this->init_sharded(page_size, num_cores, addr, command_ptr);
+    }
+
+    void init(const BufferType type, const uint32_t bank_base_address, const uint32_t page_size) {
         this->set_type(type);
         this->bank_base_address = bank_base_address;
         this->page_size_ = page_size;
+        this->sharded = false;
     }
 
+    void init_sharded(uint32_t page_size, uint32_t num_cores,  uint32_t addr, volatile tt_l1_ptr uint32_t* command_ptr){
+        this->type = BufferType::L1;
+        this->bank_base_address = addr;
+        this->page_size_ = page_size;
+        this->base_command_addr_ = command_ptr;
+        this->num_cores_ = num_cores;
+        this->sharded = true;
+
+    }
     uint32_t page_size() { return this->page_size_; }
 
-    void noc_async_write_buffer(uint32_t src, const uint32_t id, const uint32_t num_pages, const uint32_t offset) {
-        if (this->type == BufferType::SYSTEM_MEMORY) {
-            noc_async_write(src, this->get_noc_addr(id, offset), this->page_size_ * num_pages);
-        } else {
-            for (uint32_t i = 0; i < num_pages; i++) {
-                uint64_t address = this->get_noc_addr(id + i, offset);
-                noc_async_write(src, address, this->page_size_);
-                src += this->page_size_;
+
+    void noc_async_sharded_read_write_helper(const uint32_t &  addr, bool read, uint32_t num_pages, uint32_t page_id){
+        // will exit early when we have written enough
+        uint32_t num_pages_left = num_pages;
+
+        uint32_t core_id_start = 0;
+        uint32_t pages_start = 0;
+        uint32_t pages_end = 0;
+
+
+        uint32_t total_core_words = this->num_cores_ * NUM_ENTRIES_PER_SHARD;
+
+        //first get to correct core
+        for(uint32_t core_word_id = 0;  core_word_id < total_core_words; core_word_id+=NUM_ENTRIES_PER_SHARD){
+            uint32_t num_pages_core = this->base_command_addr_[core_word_id];
+            pages_end = pages_start + num_pages_core;
+            uint32_t core_id_x = this->base_command_addr_[core_word_id + 1];
+            uint32_t core_id_y = this->base_command_addr_[core_word_id + 2];
+
+            //first get to correct core
+            if(!(page_id >= pages_start && page_id < pages_end)){
+                pages_start = pages_end;
+                continue;
+            }
+            core_id_start = core_word_id;
+            break;
+        }
+
+        uint32_t flattened_page_id = page_id;
+
+        uint32_t host_page_id = 0;
+        uint32_t host_offset = 0;
+        uint32_t core_page_id = (flattened_page_id - pages_start);
+        uint32_t core_offset = core_page_id * this->page_size_;
+
+        for(uint32_t core_word_id = core_id_start;  core_word_id < total_core_words; core_word_id+=NUM_ENTRIES_PER_SHARD){
```

Comment:
Cleaner logic here would be a for loop of the following form

```
while (pages_left)
```

That way the reader can know that we just keep reading/writing pages until there are none left.